8252049: Native memory leak in ciMethodData ctor

iwanowww opened the pull request:
`ciMethodData` embeds `MethodData` to snapshot a state from the original `MethodData` instance.
But `MethodData` embeds a `Mutex` which allocates a platform-specific implementation on C-heap.
The `Mutex` is overwritten with `0`s, but the resources aren't deallocated, so the leak occurs.

Proposed fix is to run Mutex destructor right away. 

Initially, I thought about switching to `Mutex*`, but then I found that Coleen already tried that and observed a performance regression [1]. So, for now I chose the conservative approach.

In the longer term, I would consider replacing `MethodData::_extra_data_lock` with a lock-free scheme. Having a lock-per-MDO looks kind of excessive.

Testing:
- [x] verified that no memory leak observed with the reported test
- [x] tier1-4

[1] https://mail.openjdk.java.net/pipermail/hotspot-dev/2019-October/039783.html

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252049](https://bugs.openjdk.java.net/browse/JDK-8252049): Native memory leak in ciMethodData ctor


### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 011b26ca99cb211310ebe676b083368bcee6cbde


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1478/head:pull/1478`
`$ git checkout pull/1478`

iwanowww commented:
/label add hotspot-compiler
/label add hotspot-runtime
/label remove hotspot

dholmes-ora commented:
I can't say this looks pretty. It is also very hard to be able to determine when the mutex was last used and whether this running of the destructor is guaranteed to be safe. What is the lifecycle of the original MethodData?

iwanowww commented:
I agree that it's not pretty. 

I'd like to stress that the destruction happens on the freshly allocated `Mutex` and it is not related/copied from original `MethodData`:
```
MethodData::MethodData(ciMethodData& data)
  : _extra_data_lock(Mutex::leaf, "unused") {
  _extra_data_lock.~Mutex(); // release allocated resources before zeroing
```

So, I don't see any evident issues related to concurrent usage (since it shouldn't happen). Original `MethodData` lifecycle is tied to the `Method` it relates to. (But I don't see how it can matter here.) `ciMethodData` is allocated in compiler arena on per-compilarion task basis and is deallocated en masse when compilation finishes.

The alternative fix would be to introduce new ctors on `Mutex` and `os::PlatformMonitor` which omit allocation of native condition variable, but that is more intrusive IMO. Let me know what you prefer.

dholmes-ora commented:
Ah now I understand a bit better. The code creates a copy of the MethodData to preserve the data and gets the Mutex in the process. The Mutex is not wanted/needed and so it is immediately "destructed".
Okay that works.
A possible alternative might be to refactor MethodData so that the data is in MethodDataBase, and the Mutex added in MethodData, then only snapshot the MethodDataBase part in ciMethodData. But I don't know if that is feasible, nor worth the effort.

kimbarrett reviewed (approved):
I agree this fixes the immediate problem of a native memory leak.

But it all seems pretty horrible from a code understanding and maintenance
POV. So while I'm approving this change, I think there is followup work to
be done, for which an RFE should be filed.

For the specific case of ciMethodData::_orig it seems like it would be
better to separate the mutex from the data and have _orig be just the data.
That might take some significant refactoring.

But more generally, I question the embedded mutex in an MDO.

From the quoted email
(https://mail.openjdk.java.net/pipermail/hotspot-dev/2019-October/039783.html)
Coleen tried heap managing the mutex rather than having it embedded
directly, and found this caused some performace regressions.

But does there really need to be a mutex per MDO? Would a single immortal
mutex, shared by all MDOs, be adequate for this purpose? If not, maybe the
trick used by ObjectMonitor(?) -- have a set of immortal mutexes and
"randomly" choose one per MDO. This would also eliminate this part of the
walk of metadata objects to release heap resources when dropping them. And
it avoids constructing and destructing lots of mutexes (which isn't free,
even if not heap managed).

coleenp commented on src/hotspot/share/oops/methodData.cpp:
```diff
@@ -1201,6 +1202,15 @@ void MethodData::post_initialize(BytecodeStream* stream) {
+// Special ctor for ciMethodData.
+MethodData::MethodData(ciMethodData& data)
+  : _extra_data_lock(Mutex::leaf, "unused") {
+  _extra_data_lock.~Mutex(); // release allocated resources before zeroing
```
So if _extra_data_lock was a pointer to Mutex, then this statement would be simply "delete _extra_data_lock;" instead, but we'd still have this copy constructor, right?

coleenp replied:
Why isn't this a regular copy constructor taking MethodData, and pass this._orig instead and leave the zeroing in the caller?  I don't like that oops/methodData knows about ciMethodData.

iwanowww replied:
If `_extra_data_lock` were a pointer, then `MethodData::MethodData` would just `_extra_data_lock(NULL)`.

Regarding where to put zeroing, I'm fine doing it either way. 

Just want to give a try the idea David and Kim proposed with separating the lock from MDO (w/ superclass trick).

iwanowww commented:
Thanks for the feedback, David, Kim, and Coleen.

I reworked the fix and moved the fields accessed by `ciMethodData` into a separate class which is then embedded into `MethodData`. Let me know what you think about it.

coleenp reviewed (approved):
This looks like a nice refactoring to me.

iwanowww commented on src/hotspot/share/oops/methodData.hpp:
```diff
@@ -1965,23 +1968,80 @@ class MethodData : public Metadata {
     _extra_data_count   = 4     // extra DataLayout headers, for trap history
   }; // Public flag values
+
+  class MDHeader : public StackObj {
```
Fixed. 

And I was curious where `ValueObj`/`VALUE_OBJ_CLASS_SPEC` hide...  :-) Until I learned from JDK-8173070 and JDK-8198243 that it was retired for good.

coleenp replied:
yes, it's gone. thanks for making this change. thumbs up.

iwanowww commented:
> But does there really need to be a mutex per MDO? Would a single immortal
> mutex, shared by all MDOs, be adequate for this purpose? If not, maybe the
> trick used by ObjectMonitor(?) -- have a set of immortal mutexes and
> "randomly" choose one per MDO. This would also eliminate this part of the
> walk of metadata objects to release heap resources when dropping them. And
> it avoids constructing and destructing lots of mutexes (which isn't free,
> even if not heap managed).

Yes, I share the same concern about a mutex-per-MDO. Filed [JDK-8257756](https://bugs.openjdk.java.net/browse/JDK-8257756).

kimbarrett reviewed (changes requested):
Mostly good; a few small suggestions or questions.

dholmes-ora reviewed (commented):
Approval in principle from me. I'll let you work out naming details etc with Kim.

Thanks,
David

iwanowww commented:
Kim, agree on all your comments/suggestions. Let me know what you think about the latest version.

kimbarrett commented on src/hotspot/share/oops/methodData.hpp:
```diff
@@ -1983,8 +1984,8 @@ class MethodData : public Metadata {
   public:
-    MDHeader() : _nof_decompiles(0), _nof_overflow_recompiles(0), _nof_overflow_traps(0) {
-      assert(sizeof(_trap_hist) % sizeof(HeapWord) == 0, "align");
+    CompilerCounters() : _nof_decompiles(0), _nof_overflow_recompiles(0), _nof_overflow_traps(0) {
```
This doesn't initialize _creation_mileage.  Is that a bug?  Should that be a constructor argument?

iwanowww replied:
> This doesn't initialize _creation_mileage. Is that a bug? Should that be a constructor argument?

I don't think there's a bug since `MethodData` initializes it right away (and `ciMethodData` bulk copies it from the original version), but agree passing it as a constructor argument is cleaner. Updated version posted.

kimbarrett reviewed (commented):
One lingering question.  The name CompilerCounters seems much better, BTW.

iwanowww commented:
Thanks for the reviews, Kim, Coleen, and David.

iwanowww commented:
/integrate

glaubitz commented:
The change in 044616bd71ab82f0f67670152cecbabfee83d00c breaks the build on m68k which uses a native alignment of 16 bits despite being a 32-bit architecture as a result of the m68k SysV ABI.

```
In file included from /<<PKGBUILDDIR>>/src/hotspot/share/ci/ciEnv.hpp:35,
                 from /<<PKGBUILDDIR>>/src/hotspot/share/ci/compilerInterface.hpp:32,
                 from /<<PKGBUILDDIR>>/src/hotspot/share/compiler/abstractCompiler.hpp:28,
                 from /<<PKGBUILDDIR>>/src/hotspot/share/compiler/abstractCompiler.cpp:25:
/<<PKGBUILDDIR>>/src/hotspot/share/oops/methodData.hpp: In constructor 'MethodData::CompilerCounters::CompilerCounters()':
/<<PKGBUILDDIR>>/src/hotspot/share/oops/methodData.hpp:1987:55: error: static assertion failed: align
 1987 |       static_assert(sizeof(_trap_hist) % HeapWordSize == 0, "align");
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
gmake[4]: *** [lib/CompileJvm.gmk:143: /<<PKGBUILDDIR>>/build/hotspot/variant-zero/libjvm/objs/abstractCompiler.o] Error 1
gmake[3]: *** [make/Main.gmk:252: hotspot-zero-libs] Error 2
gmake[3]: *** Waiting for unfinished jobs....
```

Full log: https://buildd.debian.org/status/fetch.php?pkg=openjdk-16&arch=m68k&ver=16%7E29-1&stamp=1609359094&raw=0

Any suggestion on a quick fix (before I start digging myself)?

glaubitz commented:
> Isn't m68K a zero-only platform? In which case why are these files even part of the build let alone runtime?

Yes, it's a Zero port. But apparently the files are being used anyway.

> Exclude the assertion in a zero build?

Thanks, I'll do that. I wasn't sure whether the code was relevant for Zero or not.
